### ISSUE TITLE
Frame decoding bug in quinn-h3 (introduced during the tokio 1.0 upgrade)

### ISSUE THREAD

**Issue by djc** [quinn-rs/quinn#994]
`bench_empty_body` panics in `quinn_h3::proto::frame::PartialData::decode_data()`:

```
stack backtrace:
   0: rust_begin_unwind
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:495:5
   1: core::panicking::panic_fmt
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/panicking.rs:92:14
   2: core::panicking::panic
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/panicking.rs:50:5
   3: bytes::buf::buf_impl::Buf::copy_to_bytes
             at /Users/djc/.cargo/registry/src/github.com-1ecc6299db9ec823/bytes-1.0.1/src/buf/buf_impl.rs:816:9
   4: quinn_h3::proto::frame::PartialData::decode_data
             at ./src/proto/frame.rs:228:23
   5: <quinn_h3::frame::FrameDecoder as tokio_util::codec::decoder::Decoder>::decode::{{closure}}
             at ./src/frame.rs:56:67
   6: <quinn_h3::frame::FrameDecoder as tokio_util::codec::decoder::Decoder>::decode
             at ./src/frame.rs:56:32
   7: <tokio_util::codec::framed_impl::FramedImpl<T,U,R> as futures_core::stream::Stream>::poll_next
             at /Users/djc/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-util-0.6.2/src/codec/framed_impl.rs:137:38
   8: <tokio_util::codec::framed_read::FramedRead<T,D> as futures_core::stream::Stream>::poll_next
             at /Users/djc/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-util-0.6.2/src/codec/framed_read.rs:130:9
   9: <quinn_h3::body::RecvBody as http_body::Body>::poll_data
             at ./src/body.rs:151:33
  10: <&mut T as http_body::Body>::poll_data
             at /Users/djc/.cargo/registry/src/github.com-1ecc6299db9ec823/http-body-0.4.0/src/lib.rs:96:9
  11: quinn_h3::body::RecvBody::read_to_end::{{closure}}::{{closure}}
             at ./src/body.rs:103:40
  12: <futures_util::future::poll_fn::PollFn<F> as core::future::future::Future>::poll
             at /Users/djc/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.8/src/future/poll_fn.rs:54:9
  13: quinn_h3::body::RecvBody::read_to_end::{{closure}}
             at ./src/body.rs:102:38
  14: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /Users/djc/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  15: request::request_client::{{closure}}
             at ./benches/request.rs:186:5
```

**Comment by geieredgar** [comment]
I opened a small pull request (#995) to fix this.

**Comment by djc** [comment]
While we have landed a workaround in #995 to unblock other work, but I don't yet feel that I understand why the changes in 51685fb broke this scenario; the changes to PartialData there are pretty limited, so I wonder if the actual issue might be something else. Would be good to investigate this a bit more.
